services/api/src/utils/logging.js

```diff
@@ -1,22 +1,173 @@
-function timeStr() {
-  return new Date().toString();
+const tracer = require('./tracer');
+
+const startTime = Symbol('startTime');
+const isLive = process.env.NODE_ENV == 'production';
+
+const pinoFn = require('pino');
+
+function getGlobalContext() {
+  const currentSpan = tracer.getCurrentSpan();
+  if (!currentSpan) return null;
+  return currentSpan.context();
 }
 
-const logger = {
-  debug(...args) {
-    console.info(...[timeStr(), '-', 'DEBUG', '-'].concat(args));
-  },
-  info(...args) {
-    console.info(...[timeStr(), '-', 'INFO', '-'].concat(args));
+const parentLogger = pinoFn({
+  messageKey: 'message',
+  formatters: {
+    level(_, level) {
+      if (level === 20) {
+        return { severity: 'debug' };
+      }
+      if (level === 30) {
+        return { severity: 'info' };
+      }
+      if (level === 40) {
+        return { severity: 'warning' };
+      }
+      if (level === 50) {
+        return { severity: 'error' };
+      }
+      if (level >= 60) {
+        return { severity: 'critical' };
+      }
+      return { severity: 'default' };
+    },
   },
-  warn(...args) {
-    console.info(...[timeStr(), '-', 'WARN', '-'].concat(args));
-  },
-  error(...args) {
-    console.info(...[timeStr(), '-', 'ERROR', '-'].concat(args));
+  base: null,
+  timestamp: pinoFn.stdTimeFunctions.isoTime,
+  prettyPrint: process.env.NODE_ENV === 'dev' ? { messageKey: 'message' } : false,
+});
+
+/**
+ * A LogEntry
+ * @typedef {{request: import('http').IncomingMessage, response: import('http').ServerResponse, latency: number }} LogEntry
+ */
+
+/**
+ * @callback HttpRequestFormat
+ * @param {LogEntry} logEntry Description
+ */
+
+/**
+ * @typedef {Object} Options
+ * @property {HttpRequestFormat} [httpRequestFormat]
+ * @property {string} [useLevel=info]
+ */
+
+/**
+ *
+ * @param {*} param0
```

Comment:
is this not `Object` type? also `param0` seems a bit strange to me... not just "options"?

Reply:
yeah let me clean it up => was experimenting with jsdoc, for autocompleting, but the code has changed quite a bit since then and largely internal functions